include/circt/Dialect/RTL/RTLTypeDecls.td

```diff
@@ -0,0 +1,49 @@
+//===- RTLTypeDecls.td - RTL data type declaration ops and types ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Support for type declarations in the RTL type system.
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// Declaration operations
+//===----------------------------------------------------------------------===//
+
+def TypeScopeOp : RTLOp<"type_scope",
+      [Symbol, SymbolTable, SingleBlock, NoTerminator, NoRegionArguments,
+      HasParent<"mlir::ModuleOp">]> {
+  let summary = "Type declaration wrapper.";
+  let description = [{
+    An operation whose one body block contains type declarations. This op
+    provides a scope for type declarations at the top level of an MLIR module.
+    It is a symbol that may be looked up within the module, as well as a symbol
+    table itself, so type declarations may be looked up.
+  }];
+
+  let regions = (region SizedRegion<1>:$body);
+  let arguments = (ins SymbolNameAttr:$sym_name);
+
+  let assemblyFormat = "$sym_name $body attr-dict";
+
+  let extraClassDeclaration = [{
+    Block *getBodyBlock() { return &body().front(); }
+  }];
+}
+
+def TypedeclOp : RTLOp<"typedecl", [Symbol, HasParent<"TypeScopeOp">]> {
+  let summary = "Type declaration.";
+  let description = "Associate a symbolic name with a type.";
+
+  let arguments = (ins
+    SymbolNameAttr:$sym_name,
+    TypeAttr:$type,
+    OptionalAttr<StrAttr>:$verilogName
```

Comment:
@lattner I went with OptionalAttr here, which is in line with how extern- and generated- modules use OptionalAttr for their `verilogName` attribute. OptionalAttr also lets us elide the attribute from the printed IR form when its not set.

You initially mentioned using an empty string here, do you have a preference for that?

Reply:
I'm ok with either for now.  I suggested empty string mostly because we have been moving away from tristate optional attributes in other parts of the IR.  Instead of being able to model `None`, `""`, and `"stuff"`, they are just modeling it as `""` and `"stuff"`.  This also makes the C++ APIs less annoying to deal with.

All that said, I'm ok either way, staying consistent with the other verilogName stuff seems like a pro move to me.

Reply:
:+1: will keep the OptionalAttr for now for consistency, and we can potentially switch all the `verilogName` attributes at a later point.